OSDOCS-13509: add assembly, topic map, and concept (#92976)
openshift/openshift-docs: modules/nw-ingress-gateway-api-deployment-topologies.adoc

object restricts routes to a particular application namespace. This is the default topology for Gateway API in {product-title}.

Shared gateway:: Routes are served from multiple namespaces or with multiple hostnames. The `Gateway` object filters allow routes from application namespaces using the `spec.listeners.allowedRoutes.namespaces` field.

[id="dedicated-gateway-example_{context}"]
== Dedicated gateway example
An example of a `Gateway` resource, `fin-gateway`:

.Example dedicated `Gateway` resource
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: fin-gateway
  namespace: openshift-ingress
spec:
  listeners:
  - name: http
    protocol: HTTP
    port: 8080
----

An example of the associated `HTTPRoute` resource, `sales-db`, that attaches to the dedicated `Gateway` object:

.Example `HTTPRoute` resource
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: sales-db
  namespace: openshift-ingress
spec:
  parentRefs:
  - name: fin-gateway
  rules:
    - backendRefs:
        - name: sales-db

>         ¦ port: 8080

Author: seems to be an indentation artifact when editing yaml, it doesn't show up in the published version though

>         ¦ port: 8080

Reviewer: I can see it in the preview:
![image](https://github.com/user-attachments/assets/55dbeb85-2318-475e-ae9f-9deb9c44317e)

>         ¦ port: 8080

Reviewer: And here too:
![image](https://github.com/user-attachments/assets/cb2743c1-5bff-44f7-81bb-0533a769f0a7)

>         ¦ port: 8080

Author: it is present in the preview yeah

>         ¦ port: 8080

Author: the preview is not really reflective style-wise to the d.r.c. published documentation. even with d.o.c, the preview contained some artifacts that get stripped out during the publishing process

----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: devops-gateway
  namespace: openshift-ingress
listeners:
  - name: https
    protocol: HTTPS
    hostname: "example.com"
    allowedRoutes:
      namespaces:
        from: Selector
        selector:
        ¦ matchLabels:
        ¦   shared-gateway-access: "true"
----

Examples of the allowed namespaces for the `devops-gateway` resource:

.Example `Namespace` resources
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: dev
  labels:
    shared-gateway-access: "true"
---
apiVersion: v1
kind: Namespace
metadata:
  name: ops
  labels:
    shared-gateway-access: "true"
----

An example of two `HTTPRoute` resources, `dev-portal` and `ops-home`, that sit in different namespaces but are attached to the shared gateway:

[source,yaml]
----
apiVersion: v1
kind: HTTPRoute
metadata:
  name: dev-portal
  namespace: dev
spec:
  parentRefs:
  - name: devops-gateway
    namespace: openshift-ingress-operator
  rules:
  - backendRefs:
    - name: dev-portal
      port: 8080
---
apiVersion: v1
kind: HTTPRoute
metadata:
  name: ops-home
  namespace: ops
spec:
  parentRefs:
  - name: devops-gateway

>     namespace: openshift-ingress-operator

Reviewer: ```suggestion
    namespace: openshift-ingress
```

// Modules included in the following assemblies:
//
// * networking/gateway-api.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-ingress-gateway-api-deployment_{context}"]

> = Gateway API Deployment topologies

Author: good catch 😄 no it doesn't

> Gateway API is designed to accomodate two topologies: shared gateways or dedicated gateways. Each topology has their own advantages and as well as different security implications.

Reviewer 2: ```suggestion
Gateway API is designed to accomodate two topologies: shared gateways or dedicated gateways. Each topology has its own advantages and different security implications.
```

Dedicated gateway:: Routes and any load balancers or proxies are served from the same namespace. The `Gateway`

> object restricts routes to a particular application namespace. This is the default topology for Gateway API in {product-title}.

Author: i took that from the enhancement proposal. would `This is the default topology when deploying a Gateway API resource in {product-title} clusters` be better here you think?

> object restricts routes to a particular application namespace. This is the default topology for Gateway API in {product-title}.

Author: went ahead and updated this 😄

> Shared gateway:: Routes are served from multiple namespaces or with multiple hostnames. The `Gateway` object filters allow routes from application namespaces using the `spec.listeners.allowedRoutes.namespaces` field.

Reviewer 2: ```suggestion
Shared gateway:: Routes are served from multiple namespaces or with multiple hostnames. The `Gateway` object filters allow routes from application namespaces by using the `spec.listeners.allowedRoutes.namespaces` field.
```

[id="dedicated-gateway-example_{context}"]
== Dedicated gateway example

> An example of a `Gateway` resource, `fin-gateway`:

Reviewer 3: ```suggestion
An example of a dedicated `Gateway` resource, `fin-gateway`:
```

> An example of a `Gateway` resource, `fin-gateway`:

Reviewer 2: Consider rewriting this as a full sentence.
```suggestion
The following example shows a dedicated `Gateway` resource, `fin-gateway`:
```

.Example dedicated `Gateway` resource
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: fin-gateway
  namespace: openshift-ingress
spec:
  listeners:
  - name: http
    protocol: HTTP
    port: 8080

>     hostname: "example.com"

Reviewer 3: I think the dedicated gateway can also serve multiple httproutes in `Same` namespace, so could use wildcard hostname like `hostname: "*.example.com"`
@alebedev87 Do you think we should consider the DNSRecords creation in the example, if yes then need to set hostname like `*.dedicated.<baseDomain>`, the baseDomin could be get from `oc get dns.config cluster -ojsonpath="{.spec.baseDomain}"`

Author: this sounds like it can be an continuous improvement task for a new PR. the current one is pretty large and has had enough scope creep already

----

> An example of the associated `HTTPRoute` resource, `sales-db`, that attaches to the dedicated `Gateway` object:

Reviewer 2: ```suggestion
The following example shows the associated `HTTPRoute` resource, `sales-db`, which attaches to the dedicated `Gateway` object:
```

.Example `HTTPRoute` resource
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: sales-db

>   namespace: openshift-ingress

Reviewer 3: In this example the HTTPRoute `sales-db` is in `openshift-ingress` namespace, but it might not be a case in practice. We mentioned Dedicated gateway previously that  
```
The `Gateway` object restricts routes to a particular application namespace.
```
Maybe we could add namespaces selector to gateway and label to namespace as well (similar to shared gateway but just apply to one dedicated namespace) ?

spec:
  parentRefs:
  - name: fin-gateway
  rules:
    - backendRefs:
        - name: sales-db
        ¦ port: 8080
----

> The `HTTPRoute` resource must have the `Gateway` object's name as the value for its `parentRefs` field in order to attach to the gateway. Implicitly, the route is assumed to be in the same namespace as the `Gateway` object.

Reviewer 2: ```suggestion
The `HTTPRoute` resource must have the name of the `Gateway` object as the value for its `parentRefs` field in order to attach to the gateway. Implicitly, the route is assumed to be in the same namespace as the `Gateway` object.
```

[id="shared-gateway-example_{context}"]
== Shared gateway example

> An example of a `Gateway` resource, `devops-gateway`, that has a label selector `spec.listeners.allowedRoutes.namespaces` set to match any namespaces containing `shared-gateway-access: "true"`:

Reviewer 2: ```suggestion
The following example shows a `Gateway` resource, `devops-gateway`, that has a `spec.listeners.allowedRoutes.namespaces` label selector set to match any namespaces containing `shared-gateway-access: "true"`:
```

.Example shared `Gateway` resource
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: devops-gateway
  namespace: openshift-ingress
listeners:
  - name: https
    protocol: HTTPS
    hostname: "example.com"
    allowedRoutes:
      namespaces:
        from: Selector
        selector:
        ¦ matchLabels:
        ¦   shared-gateway-access: "true"
----

> Examples of the allowed namespaces for the `devops-gateway` resource:

Reviewer 2: ```suggestion
The following examples show the allowed namespaces for the `devops-gateway` resource:
```

.Example `Namespace` resources
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: dev
  labels:
    shared-gateway-access: "true"
---
apiVersion: v1
kind: Namespace
metadata:
  name: ops
  labels:
    shared-gateway-access: "true"
----

> An example of two `HTTPRoute` resources, `dev-portal` and `ops-home`, that sit in different namespaces but are attached to the shared gateway:

Reviewer 2: ```suggestion
In this example, two `HTTPRoute` resources, `dev-portal` and `ops-home`, are in different namespaces but are attached to the shared gateway:
```

[source,yaml]
----
apiVersion: v1
kind: HTTPRoute
metadata:
  name: dev-portal
  namespace: dev
spec:
  parentRefs:
  - name: devops-gateway
    namespace: openshift-ingress
  rules:
  - backendRefs:
    - name: dev-portal
      port: 8080
---
apiVersion: v1
kind: HTTPRoute
metadata:
  name: ops-home
  namespace: ops
spec:
  parentRefs:
  - name: devops-gateway
    namespace: openshift-ingress
  rules:
  - backendRefs:
    - name: ops-home
      port: 8080
----

> With a shared gateway topology, the routes must specify the namespace of the `Gateway` object it wants to attach to. Multiple `Gateway` objects can be deployed and shared across namespaces. When there are multiple shared gateways this becomes conceptually similar to Ingress Controller sharding.

Reviewer 2: ```suggestion
With a shared gateway topology, the routes must specify the namespace of the `Gateway` object it wants to attach to. Multiple `Gateway` objects can be deployed and shared across namespaces. When there are multiple shared gateways, this becomes conceptually similar to Ingress Controller sharding.
```
It might help to add a noun after "this" in the last sentence. Should it be "When there are multiple shared gateways, this topology becomes conceptually similar to Ingress Controller sharding."?
